text/0000-cargo-embed-dependency-versions.md

```diff
@@ -0,0 +1,195 @@
+[motivation]: #motivation
+
+Rust is very promising for security-critical applications due to its safety guarantees, but there currently are gaps in the ecosystem that prevent it. One of them is the lack of any infrastructure for security updates.
+
+Linux distributions alert you if you're running a vulnerable software version and you can opt in to automatic security updates. Cargo not only has no automatic update infrastructure, it doesn't even know which libraries or library versions went into compiling a certain binary, so there's no way to check if your system is vulnerable or not.
+
+The primary use case for this information is cross-referencing versions of the dependencies against [RustSec advisory database](https://github.com/RustSec/advisory-db) and/or third-party databases such as [Common Vulnerabilities and Exposures](https://en.wikipedia.org/wiki/Common_Vulnerabilities_and_Exposures). This also enables use cases such as ensuring a fix in a depencency has been propagated across the entirety of your fleet or preventing binaries with unvetted dependencies from accidentally reaching a production environment - all with zero bookkeeping.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Every time an executable is compiled with Cargo, the dependency tree of the executable is recorded in the binary. This includes the names, versions, dependency kind (build or runtime), and origin kind (crates.io, git, local filesystem, custom registry). Development dependencies are not recorded, since they cannot affect the final binary. All filesystem paths and URLs are redacted to preserve privacy. The data is encoded in JSON and compressed with zlib to reduce its size.
+
+This data can be recovered using existing tools like `readelf` or Rust-specific tooling. It can be then used to create a Software Bill of Materials in a common format, or audit the dependency list for known vulnerabilities.
+
+WASM, asm.js and embedded platforms are exempt from this mechanism by default since they have very strict code size requirements. For those platforms we encourage you to use tooling that record the hash of every executable in a database and associates the hash with its Cargo.lock, compiler and LLVM version used for the build.
+
+A configuration option can be used to opt out of this behavior if it is not desired (e.g. when building [extremely minimal binaries](https://github.com/johnthagen/min-sized-rust)).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The version information is encoded in an additional arbitrary section of the executable by Cargo. The exact mechanism varies depending on the executable format (ELF, Mach-O, PE, etc.). The section name is `.dep-v0` across all platforms. The section name must be changed if breaking changes are made to the format.
+
+The data is encoded in JSON which is compressed with Zlib. All arrays a sorted not to disrupt reproducible builds.
```

Comment:
What is the build-time impact of this change?

I know first-party will be faster but still wondering about the overhead.  Both for debug builds generally but also with the upcoming cargo-script

I wonder if we should enable it on a per-profile basis so it can be enabled for release builds but not debug builds

Reply:
[`cargo-binstall` CI](https://github.com/cargo-bins/cargo-binstall/actions/runs/5758584682/job/15611362302) has been using `cargo-auditable` for release build and it seems that impact on build time is insignificant.

Reply:
You can measure the impact right now with this command:

```bash
hyperfine --max-runs 5 --prepare 'cargo clean' 'cargo build' 'cargo auditable build'
```

It shows a 3% build time regression when building `cargo auditable` itself in debug mode, or a 400ms increase. A large portion of that time is spent in a call to `cargo metadata` and repeated calls to `rustc -vV`. (This is on a system with `rustup`, so most of that time is spent in `rustup` parsing its configuration on every call to `rustc`). 

`cargo auditable` itself is so fast that it doesn't even show up on the profile with a 1000Hz sampling interval: https://share.firefox.dev/4529Yq8

The code is also quite naive right now. I have not attempted to optimize it in any way because it was already fast enough, so there are performance gains to be had if build time ever becomes a problem. The only thing I have tried using lighter Zlib compression in debug mode builds, but this did not result in any measurable changes because the files being compressed are just dozens of kilobytes and that's near-instant regardless.

Reply:
On a [large workspace](https://github.com/rustsec/rustsec) `cargo auditable` actually comes out 5% faster on my machine, so the compile time differences must be within noise threshold.

Reply:
Thanks! And to be clear, I expect this to not be too noticeable for most projects but I am protective of build times for "cargo script".

Reply:
I still wonder if we should distinguish between debug and release builds.

Reply:
> Thanks! And to be clear, I expect this to not be too noticeable for most projects but I am protective of build times for "cargo script".

I think it does not make sense to enable this for "cargo script", given that it is distributed by source code instead of the pre-built artifacts?

Reply:
I don't think it's possible to decide at the design stage, without trying it and seeing what the impact actually is.

I can add a sentence about it being always enabled for release builds, and for debug builds only if the compilation time impact is deemed to be negligible, how does that sound?

Reply:
Id rather not hack up cargo with special cases for "cargo script" and we do plan to eventually allow publishing them which means you can `cargo install` one, one day.

With that long term route in mind, we shouldn't preclude "cargo scripts" from using this feature. However, they are run in debug mode by default, so if we make put the setting in the profile and have it be different for debug vs release, then day-to-day "cargo scripts" won't use it.

Reply:
I have included the build time considerations into the RFC, and added a clause that if the compile time impact ends up being noticeable, this feature will be disabled for debug build by default prior to stabilization.